File: leetcode_problems/cpp/lc1718_construct_lexic_largest_valid_sequence.cpp
#include <vector>
using std::vector;

class Solution
{
public:
    bool large(vector<int> seqA, vector<int> seqB)
    {
        if (seqA.size() > seqB.size())
        {
            return true;
        }
        else if (seqA.size() < seqB.size())
        {
            return false;
        }
        else
        {
            for (int i = 0; i < seqA.size(); i++)
            {
                if (seqA[i] == seqB[i])
                {
                    continue;
                }
                return seqA[i] > seqB[i];
            }
            return false;
        }
    }

    void backtrack(
        int i,
        int n,
        vector<bool> &nums,
        vector<int> &stack,
        vector<int> &res,
        bool &found)
    {
        if (found)
        {
            return;
        }

        if (i >= stack.size())
        {
            if (large(stack, res))
            {
                res = stack;
                found = true;
            }
            return;
        }

        if (stack[i] != -1)
        {
            backtrack(i + 1, n, nums, stack, res, found);
            return;
        }

        for (int num = n; num > 0; num--)
        {
            if (!nums[num])
            {
                continue;
            }

            int dist = num == 1 ? 0 : num;
            if (i + dist >= stack.size() || stack[i + dist] != -1)
            {
                continue;
            }

            stack[i] = num;
            stack[i + dist] = num;
            nums[num] = false;

            backtrack(i + 1, n, nums, stack, res, found);

            stack[i] = -1;
            stack[i + dist] = -1;
            nums[num] = true;
        }
    }

    vector<int> constructDistancedSequence(int n)
    {
        bool found = false;
        vector<bool> nums(n + 1, true);
        vector<int> stack(1 + (n - 1) * 2, -1);
        vector<int> res;
        backtrack(0, n, nums, stack, res, found);
        return res;
    }
};

File: leetcode_problems/python/lc1718_construct_lexic_largest_valid_sequence.py
class Solution:
    def constructDistancedSequence(self, n: int) -> list[int]:
        def large(seqA: list[int], seqB: list[int]) -> bool:
            if len(seqA) > len(seqB):
                return True
            elif len(seqA) < len(seqB):
                return False
            else:
                for i in range(len(seqA)):
                    if seqA[i] == seqB[i]:
                        continue
                    return seqA[i] > seqB[i]
                return False

        def backtrack(i: int) -> None:
            nonlocal stack
            nonlocal res
            nonlocal found

            if found:
                return

            if i >= len(stack):
                if large(stack, res):
                    res = stack[:]
                    found = True
                return None

            if stack[i] != -1:
                backtrack(i + 1)
                return None

            for num in range(n, 0, -1):
                if not nums[num]:
                    continue

                dist = 0 if num == 1 else num
                if i + dist >= len(stack) or stack[i + dist] != -1:
                    continue

                stack[i] = num
                stack[i + dist] = num
                nums[num] = False

                backtrack(i + 1)

                nums[num] = True
                stack[i] = -1
                stack[i + dist] = -1

        found = False
        nums = [True] * (n + 1)
        stack = [-1] * (1 + (n - 1) * 2)
        res = []
        backtrack(0)
        return res


if __name__ == "__main__":
    n = 3
    print(Solution().constructDistancedSequence(n))
